File: dairytrack-api/MilkProduction/app/models/cow.py
from app import db

class Cow(db.Model):
    __tablename__ = 'cows'

    id = db.Column(db.Integer, primary_key=True)
    farmer_id = db.Column(db.Integer, db.ForeignKey('farmers.id'), nullable=False)
    name = db.Column(db.String(50))
    breed = db.Column(db.String(50))
    birth_date = db.Column(db.Date)
    lactation_status = db.Column(db.Boolean, default=False)
    lactation_phase = db.Column(db.String(20), default=None)
    weight_kg = db.Column(db.Numeric(5, 2))
    reproductive_status = db.Column(db.String(20))
    gender = db.Column(db.String(10))
    entry_date = db.Column(db.Date)
    notifications = db.relationship('Notification', back_populates='cow')
    created_at = db.Column(db.DateTime, default=db.func.current_timestamp())
    updated_at = db.Column(db.DateTime, default=db.func.current_timestamp(), onupdate=db.func.current_timestamp())
    
    # Relationship with RawMilk
    raw_milks = db.relationship('RawMilk', back_populates='cow', cascade='all, delete-orphan')

    def to_dict(self, include_raw_milks=True):
        return {
            'id': self.id,
            'farmer_id': self.farmer_id,
            'name': self.name,
            'breed': self.breed,
            'birth_date': self.birth_date.strftime('%Y-%m-%d'),
            'lactation_status': self.lactation_status,
            'lactation_phase': self.lactation_phase,
            'weight_kg': str(self.weight_kg),
            'reproductive_status': self.reproductive_status,
            'gender': self.gender,
            'entry_date': self.entry_date.strftime('%Y-%m-%d'),
            'notifications': [notification.to_dict() for notification in self.notifications],
            'raw_milks': [raw_milk.to_dict(include_cow=False) for raw_milk in self.raw_milks] if include_raw_milks else []
        }

    def __repr__(self):
        return f"Cow('{self.name}', '{self.breed}')"

File: dairytrack-mobile/lib/model/peternakan/cow.dart
class Cow {
  int? id;
  int? farmerId;
  final String name;
  final String breed;
  final DateTime birthDate;
  final bool lactationStatus;
  final String? lactationPhase;
  final double weight_kg;
  final String reproductiveStatus;
  final String gender;
  final DateTime entryDate;
  final DateTime createdAt;
  final DateTime updatedAt;

  Cow({
    this.id,
    this.farmerId,
    this.name = 'Unknown', // Default value
    this.breed = 'Unknown', // Default value
    DateTime? birthDate, // Nullable
    this.lactationStatus = false, // Default value
    this.lactationPhase,
    this.weight_kg = 0.0, // Default value
    this.reproductiveStatus = 'Unknown', // Default value
    this.gender = 'Unknown', // Default value
    DateTime? entryDate, // Nullable
    DateTime? createdAt, // Nullable
    DateTime? updatedAt, // Nullable
  })  : birthDate = birthDate ?? DateTime(1970, 1, 1), // Default value
        entryDate = entryDate ?? DateTime(1970, 1, 1), // Default value
        createdAt = createdAt ?? DateTime(1970, 1, 1), // Default value
        updatedAt = updatedAt ?? DateTime(1970, 1, 1); // Default value

  factory Cow.fromJson(Map<String, dynamic> json) {
    return Cow(
      id: json['id'],
      farmerId: json['farmer_id'],
      name: json['name'] ?? 'Unknown',
      breed: json['breed'] ?? 'Unknown',
      birthDate: json['birth_date'] != null
          ? DateTime.parse(json['birth_date'])
          : null,
      lactationStatus: json['lactation_status'] ?? false,
      lactationPhase: json['lactation_phase'],
      weight_kg: json['weight_kg'] != null
          ? double.tryParse(json['weight_kg'].toString()) ?? 0.0
          : 0.0,
      reproductiveStatus: json['reproductive_status'] ?? 'Unknown',
      gender: json['gender'] ?? 'Unknown',
      entryDate: json['entry_date'] != null
          ? DateTime.parse(json['entry_date'])
          : null,
      createdAt: json['created_at'] != null
          ? DateTime.parse(json['created_at'])
          : null,
      updatedAt: json['updated_at'] != null
          ? DateTime.parse(json['updated_at'])
          : null,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      if (id != null) 'id': id,
      if (farmerId != null) 'farmer_id': farmerId,
      'name': name,
      'breed': breed,
      'birth_date': birthDate.toIso8601String(),
      'lactation_status': lactationStatus,
      'lactation_phase': lactationPhase,
      'weight_kg': weight_kg,
      'reproductive_status': reproductiveStatus,
      'gender': gender,
      'entry_date': entryDate.toIso8601String(),
      'created_at': createdAt.toIso8601String(),
      'updated_at': updatedAt.toIso8601String(),
    };
  }
}
